src/engine/strat_engine/range_alloc.rs

```diff
@@ -57,44 +97,58 @@ impl RangeAllocator {
         for &(off, len) in to_free {
             self.check_for_overflow(off, len);
 
-            for val in *off..*off + *len {
-                let removed = self.used.remove(&val);
-                if removed == false {
-                    panic!(format!("tried to remove value {} not present in RangedSet", val));
+            let (prev_off, prev_len) = self.used
+                .range((Unbounded, Included(off)))
+                .rev()
+                .next()
+                .map(|(k, v)| (*k, *v))
+                .expect("must match an existing range");
```

Comment:
Why is ```remove_ranges()``` public? This affects how you should deal with it. If it were private, then public methods would control how it was invoked. But since it's public, different rules. I realize that at this point, nothing is invoking it.

Speaking of which, I just noticed that ```new_with_used``` can also panic if its ranges overlap. Given that it is public, that is kind of wrong, so something should be done. The most correct thing would be for ```insert_ranges()``` to return a Result and for ```new_with_used``` to pass on that result, I think.

Difference between expect() and a straight up panic!() is that expect means that you have demonstrated by an argument that the unwrap must always succeed. panic() means you haven't, but you're not going to handle the error, regardless. But given that the method in question is public, switching from expect to panic may not be the right thing, anyway.